eclipse-theia/theia
# Support TestMessageStackFrame API

#### What it does
Adds 'stacktraces: TestMessageStackFrame' to TestMessage. This optional property allows extensions to describe the context in which an error could occur when test is failing. 
Here is a snapshot as an example on vscode. They embed small monaco editors to display the line and a bit around to display the stack trace when the test is failing.

![test-stack](https://github.com/user-attachments/assets/4f4e04e2-33ef-4c9d-b274-6e4589646f54)

Note that our current implementation of Test is lighter than the vs code one. So a more simple approach was given, with only the label of the StackTrace and a hyperlink to navigate to the file located at the uri given by the stack frame, as well as the cursor position.

![image](https://github.com/user-attachments/assets/fd5fb495-64c9-4215-88b9-345e190e89f8)

Fixes #14111

Contributed on behalf of STMicroelectronics

#### How to test
The attached extension is based on test-sample-provider from vscode-extension-samples:
- zipped vsix: [test-provider-sample-0.0.1.zip](https://github.com/user-attachments/files/16965212/test-provider-sample-0.0.1.zip)
- src: [test-provider-sample-0.0.1-src.zip](https://github.com/user-attachments/files/16965219/test-provider-sample-0.0.1-src.zip)

1. The testing extension will create tests for each mathematical expressions located in markdown files. It will also randomly add 0 to 3 stack traces to the TestMessage when the test is in failure, to test the cases with 0, 1 or more stack frames in test message.
2. Have several md files in the workspace, with some computation in it. For my testing, I used the following set of files:
- [test.zip](https://github.com/user-attachments/files/16965164/test.zip)
3. Runs some of the tests from test explorer, with some ones failing. Ensure that the messages are displayed in the Test Results view and that the links can be navigated to the files in the workspace. The cursor shall be placed in the right location. 
 
#### Follow-ups

None known

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
